src/taskgraph/generator.py

```diff
@@ -31,22 +31,47 @@ class KindNotFound(Exception):
     """
 
 
+# In an ideal world these would be part of each loader. Unfortunately, loaders
+# can only provide information as part of yielding task inputs, and we need to
+# know the set of transforms prior to starting to read from the loader.
+DEFAULT_TRANSFORMS = [
+    "taskgraph.transforms.job:transforms",
+    "taskgraph.transforms.task:transforms",
+]
+
+
 @attr.s(frozen=True)
 class Kind:
     name = attr.ib(type=AnyStr)
     path = attr.ib(type=AnyStr)
     config = attr.ib(type=dict)
     graph_config = attr.ib(type=GraphConfig)
 
-    def _get_loader(self):
+    def _get_loader_and_transforms(self):
+        transform_refs = copy.copy(self.config.get("transforms", []))
         try:
             loader = self.config["loader"]
         except KeyError:
-            raise KeyError(f"{self.path!r} does not define `loader`")
-        return find_object(loader)
+            # To avoid confusion, we do not allow these transforms to be
+            # specified if they are going to be added automatically. This
+            # avoids issues such as including them twice, being in a different
+            # order in the kind vs. here, etc.
+            for t in DEFAULT_TRANSFORMS:
+                if t in transform_refs:
+                    raise KeyError(
+                        "neither job nor task transforms may be specified unless you also specify a loader"
+                    )
+            loader = "taskgraph.loader.transform:loader"
+            transform_refs.extend(DEFAULT_TRANSFORMS)
```

Comment:
It seems a bit surprising to me that we add the default transforms if you don't specify a loader in kind.yml, but not if you set it, even to the default.  I guess it sort of makes sense because it lets you keep backwards compat with existing kind.yml files, but conceptually it feels odd.

Reply:
I agree with you, actually, and I'm open to change here. A couple of things I experimented with were:
* A new loader called `default` that has these default transforms set. (The actual implementation just called out to the `transforms` loader - so it was really just a way to uniquely identify the "default" case - otherwise it functionally identical to the `transform` loader.) The main reason I didn't end up with this is because it felt like overcomplicating things.
* Adding the concept of default transforms to loaders in general, by putting that list in, eg: `loader/transform.py`. I didn't pursue this in the end because I wasn't sure how to handle the case of conflicts between the `kind` and the defaults in the loaders. (Perhaps this is the same thing we're discussing about the current patch though...)

Reply:
Tbh I like the `default` loader idea the best and feel like it's less complicated than the current implementation. The main reason is that the current implementation is making the false assumption that loaders have transforms. But in reality, one could implement a loader that doesn't use transforms at all (I believe avoiding transforms was even the original intent of loaders)! Personally I feel like loaders are too complicated and should be removed, but as long as we have them, I'd prefer to try and avoid assuming they are transform based.

Reply:
I'm not even sure how you get away without at least the `task` transform in this day and age :upside_down_face: - but you're right, this is definitely an assumption that isn't necessarily true.

I'll take another pass on the `default` idea.

Reply:
Something I noticed on a read -- it's actually the case right now that [we require them to have at least an empty array of transforms defined](https://github.com/taskcluster/taskgraph/blob/3018cf24e697b60ad067c276ac03ee046c63f6b8/src/taskgraph/generator.py#L59) :grimacing:.

And one more clarification: for the new, `default` loader, we clearly must assume that it uses transforms in order to set these defaults -- right? Is there a preference between leaving them in `generator.py` as they are now, or tossing them in the new `taskgraph.loader.default` module? (The latter may imply that all loaders support default transforms...which it sounds like we don't want to do?)

Reply:
TIL! I feel less strongly about introducing further assumptions then.

I think they'd be better in `taskgraph.loader.default` because that is what is consuming it. To me `generator.py` feels like the wrong place.

Reply:
Yeah, I agree that it feels like the better place for it. Given that though - should be looking in _all_ loaders for `DEFAULT_TRANSFORMS`, then? Or is it a special behaviour only for the `default` one? (Presumably we'd look for it regardless of whether the `default` loader was explicitly specified or not, too.)

Reply:
Oh I see, I guess I was thinking the implementation logic could also go in `loader/default.py`, but I guess it needs to be in `generator.py` for some reason?

I guess it makes sense to look for it in all loaders in that case? I dunno.

Reply:
Right -- it can't go in the `loader` function because that function doesn't yield anything until transforms begin yielding. We need to know which transforms are being used prior to running any existing loading code.

We could certainly add a new function that we look for in a loader module that returns default transforms, but anything we do that puts this proposed set of defaults into `loader/default.py` sortof implies that it's something that could be used by any loader.

Reply:
([Our call to the loader function](https://github.com/taskcluster/taskgraph/blob/main/src/taskgraph/generator.py#L57 return) returns a generator, and the yielding doesn't happen until [we start running transforms](https://github.com/taskcluster/taskgraph/blob/main/src/taskgraph/generator.py#L86). I don't _think_ there's a way to have it, eg: return a tuple of generator + transforms, although perhaps that's wrong.)

Reply:
The loader function could e.g. modify `config["transforms"]` and then return the generator, instead of yielding things directly (as in https://github.com/jcristau/taskgraph/commit/2c420c06738c4fd30040054c5531cbb5a5ad85b4)?
Or just return a generator + transforms, indeed (as in https://github.com/jcristau/taskgraph/commit/6da7c2123ea13b6017d4772f7f5b64d048b28f78).
(I'm not saying that's necessarily better, e.g. I'm not a fan of the side effect in the first patch, or the backwards compat handling in the second one, but if we want to keep the logic to the default loader instead of the generator, it's a possibility)